vignettes/tutorial/cypress.Rmd

```diff
@@ -0,0 +1,568 @@
+  beforeEach(() => {
+    cy.visit("/");
+  });
+
+  it("'Show message' button exists", () => {
+    cy.get(".message button").should("have.text", "Show message");
+  });
+
+  it("'Show message' button shows the message'", () => {
+    cy.get(".message button").click();
+  });
+});
+
+```
+
+Now, we just need to check if the message is there and has proper content.
+We will do it the same as we did with the button - we will look for it using
+the CSS selector and check if it contains the expected text.
+This time we will use the id. If you check the element in the browser
+(the same way as we did with the button), you will see that
+the id of the message is "app-message-message_text". This shows, that in Rhino,
+everything is a Shiny module, even the outermost UI and server part.
+"app" in this selector is the namespace that identifies this outer module.
+"message" is the namespace part added when we called our `message.R` module
+and "message_text" is the id of our `textOutput`.
+
+Ok, let's use it and check if the message element has a correct text inside:
+
+```js
+// tests/cypress/integration/message.js
+
+describe("Show message", () => {
+  beforeEach(() => {
+    cy.visit("/");
+  });
+
+  it("'Show message' button exists", () => {
+    cy.get(".message button").should("have.text", "Show message");
+  });
+
+  it("'Show message' button shows the message'", () => {
+    cy.get(".message button").click();
+
+    cy.get("#app-message-message_text").should("have.text", "This is a message");
+  });
+});
+
+```
+
+Run tests again and you should see that again, all passed:
+
+![](images/message_e2e_test_2.png)
+
+# Test counting clicks functionality
+
+As previously, we need to start with creating a file and providing the test structure.
+Create `tests/cypress/integration/clicks.js` and fill it with the skeleton
+similar to the first test file:
+
+```js
+// tests/cypress/integration/clicks.js
+
+describe("Counting clicks", () => {
+  beforeEach(() => {
+    cy.visit("/");
+  });
+
+  it("has a 'Click me!' button", () => {
+
+  });
+
+  it("counts clicks", () => {
+
+  });
+});
+```
+
+## Test if the button exists
+
+First, we want to make sure the button exists. We will follow the example from
+the message module:
+
+```js
+// tests/cypress/integration/clicks.js
+
+describe("Counting clicks", () => {
+  beforeEach(() => {
+    cy.visit("/");
+  });
+
+  it("has a 'Click me!' button", () => {
+    cy.get(".clicks button").should("have.text", "Click me!");
+  });
+
+  it("counts clicks", () => {
+
+  });
+});
+```
+
+Now, if you run tests, you will see the new ones. And they should pass.
+
+![](images/clicks_e2e_test_1.png)
+
+## Test if clicks are counted correctly
+
+The next step is to check if clicking the button affects the counter
+and if it shows the correct value. To do that, we will simulate 5 clicks
+and check if the counter is set to 5.
+
+Since the button will be clicked several times, we need to assign it to a variable.
```

Comment:
Technically it's called an alias; [variables can work in somewhat unexpected way](https://docs.cypress.io/guides/core-concepts/variables-and-aliases).